compiler/rustc_hir_analysis/src/collect/predicates_of.rs

```diff
@@ -618,45 +627,28 @@ pub(super) fn implied_predicates_with_filter(
     let icx = ItemCtxt::new(tcx, trait_def_id);
 
     let self_param_ty = tcx.types.self_param;
-    let (superbounds, where_bounds_that_match) = match filter {
-        PredicateFilter::All => (
-            // Convert the bounds that follow the colon (or equal in trait aliases)
-            icx.astconv().compute_bounds(self_param_ty, bounds, OnlySelfBounds(false)),
-            // Also include all where clause bounds
-            icx.type_parameter_bounds_in_generics(
-                generics,
-                item.owner_id.def_id,
-                self_param_ty,
-                OnlySelfBounds(false),
-                None,
-            ),
-        ),
-        PredicateFilter::SelfOnly => (
-            // Convert the bounds that follow the colon (or equal in trait aliases)
-            icx.astconv().compute_bounds(self_param_ty, bounds, OnlySelfBounds(true)),
-            // Include where clause bounds for `Self`
-            icx.type_parameter_bounds_in_generics(
-                generics,
-                item.owner_id.def_id,
-                self_param_ty,
-                OnlySelfBounds(true),
-                None,
-            ),
-        ),
-        PredicateFilter::SelfThatDefines(assoc_name) => (
-            // Convert the bounds that follow the colon (or equal) that reference the associated name
-            icx.astconv().compute_bounds_that_match_assoc_item(self_param_ty, bounds, assoc_name),
-            // Include where clause bounds for `Self` that reference the associated name
-            icx.type_parameter_bounds_in_generics(
-                generics,
-                item.owner_id.def_id,
-                self_param_ty,
-                OnlySelfBounds(true),
-                Some(assoc_name),
-            ),
-        ),
+    let superbounds = match filter {
+        // Should imply both "real" supertraits, and also associated type bounds
+        // from the supertraits position.
+        PredicateFilter::All | PredicateFilter::SelfAndAssociatedTypeBounds => {
+            icx.astconv().compute_bounds(self_param_ty, bounds, OnlySelfBounds(false))
+        }
+        // Should only imply "real" supertraits, i.e. predicates with the self type `Self`.
+        PredicateFilter::SelfOnly => {
+            icx.astconv().compute_bounds(self_param_ty, bounds, OnlySelfBounds(true))
+        }
+        PredicateFilter::SelfThatDefines(assoc_name) => {
+            icx.astconv().compute_bounds_that_match_assoc_item(self_param_ty, bounds, assoc_name)
+        }
     };
```

Comment:
This is getting complex enough that I wonder if we should simplify. Some thoughts:
- Any way we can merge `compute_bounds` and `compute_bounds_that_match_assoc_items`?
- Should `compute_bounds` take a `PredicateFilter`?

Reply:
I think so, passing the whole predicate filter to compute_bounds allows us to merge the two fns. I'll give it a shot.